File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
type OptNode = Option<Rc<RefCell<TreeNode>>>;
impl Solution {
    pub fn min_depth(root: OptNode) -> i32 {
        fn depth(root: &OptNode) -> i32 {
            match root {
                Some(node) => {
                    let node = node.borrow();
                    if node.left.is_none() && node.right.is_none() {
                        1
                    } else {
                        1 + depth(&node.left).min(depth(&node.right))
                    }
                }
                None => i32::MAX,
            }
        }


        match root {
            Some(_) => depth(&root),
            None => 0
        }
    }
}

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
        if(!root) return 0;
    // if(root.right !== null && root.left !== null){
    //     return 1 + Math.min(minDepth(root.left), minDepth(root.right));
    // } else {
    //     return 1 + Math.max(minDepth(root.left), minDepth(root.right));
    // }

    let queue = [root];
    let height = 0;

    while(queue.length > 0){
        let len = queue.length;
        height++;
        for(let i = 0; i < len; i++){
            let node = queue.shift();
            if(!node.left && !node.right) return height;
            if(node.left) queue.push(node.left)
            if(node.right) queue.push(node.right)
        }
    }
};